app/controllers/groups_controller.rb

```diff
@@ -0,0 +1,13 @@
+class GroupsController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    raise ActionController::RoutingError, 'Not Found' if @groups.blank?
+  end
+
+  def show
+    @group = Group.find(params[:id])
+    raise ActionController::RoutingError, 'Not Found' unless @groups.include?(@group)
```

Comment:
If `@group` must be in `@groups` you can simplify this and avoid a trip to the DB:

```ruby
@group = @groups.find { |group| group.id == params[:id] }
```